### DISCUSSION TITLE
OpenDTU as PV Inverter in Victron Venus OS

### DISCUSSION THREAD

**Discussion by henne49** [tbnobody/OpenDTU#349] [Show and tell]
inspired by the dbus-shelly-3em-smartmeter from [@fabian-lauer](https://community.victronenergy.com/users/24700/fabian-lauer.html), and dbus-shelly-1pm-pvinverter from [@vikt0rm](https://community.victronenergy.com/users/22329/vikt0rm.html), I have created a Github project to usen OpenDTU for Hoymiles and TSUN Inverter API to show PV values in VenusOS as a PV inverter. So you do not need any Shelly or metering device to get a reading into Venus and VRM Portal. Hope you find this useful. Attached are some screenshots and the link to Github: https://github.com/henne49/dbus-opendtu/tree/main

https://community.victronenergy.com/questions/169076/opendtu-as-pv-inverter-in-venusos.html
https://github.com/henne49/dbus-opendtu 

<img width="482" alt="devicelist" src="https://user-images.githubusercontent.com/4662326/201899755-46ee63f8-3e1a-4b46-8ab3-a00b5ef27db0.png">
<img width="481" alt="device" src="https://user-images.githubusercontent.com/4662326/201899764-a50241ac-3b47-4429-9762-866515d3ba92.png">
<img width="483" alt="overview" src="https://user-images.githubusercontent.com/4662326/201899780-0028f2eb-a05c-4f00-91f6-bf248f995e36.png">
<img width="481" alt="devicedetails" src="https://user-images.githubusercontent.com/4662326/201899782-6b6ea6ee-de3a-4316-85bb-392ce44378c4.png">

**Comment by stromibaer** [comment]
The scripts has to be installed on the GX device and the GX device gets the values via the API interface from the Open-DTU?

I have extende the Open-DTU code like the Home Assistant code to register the inverters as Mqtt device with:
https://github.com/freakent/dbus-mqtt-devices
and sending values via mqtt to the gx device. The inverters can be configured in Open-DTU on which current phase they are working and in the GX device if they are on AC-IN or AC-OUT side. 

Code is fork from OpenDTU:
https://github.com/stromibaer/OpenDTU-Victron

![image](https://user-images.githubusercontent.com/47776842/202280921-32a1c8b0-8abe-4032-a4be-19f8ca94686e.png)

![image](https://user-images.githubusercontent.com/47776842/202281096-6a46b516-36c1-401c-87eb-846c45428bc1.png)

![image](https://user-images.githubusercontent.com/47776842/202281239-7fbbbf59-a0d8-4d54-b202-aeb309c5cbab.png)

![image](https://user-images.githubusercontent.com/47776842/202281326-0a21fc39-78c8-4020-883d-d3b8b929377f.png)


![image](https://user-images.githubusercontent.com/47776842/202282112-c9a8afae-9630-4455-aae6-d087b7c8c391.png)

![image](https://user-images.githubusercontent.com/47776842/202282020-7c7be444-1b75-471a-890f-88f9c56a2946.png)

Code is running currently in 2 Victron setups for long time tests and i will prepare pull request to OpenDTU project.
Maybe i find some more tester.

**Comment by henne49** [comment]
Correct, my script has to be installed on the GX Device and is running there, I like your approach as well to it doing the job diretcly from openDTU. I think your repo is private, I cannot access your code.

**Reply by stromibaer** [reply]
Pls try the link again, maybe it was not formatted the right way.

**Reply by henne49** [reply]
now it works, would you mind sharing the binary, then I can test it tomorrow and see how it works.

**Reply by stromibaer** [reply]
What kind of ESP32 are you using? 
The normal one or Olimex or ...?

**Reply by henne49** [reply]
Node MCU ESP32 38Pin
https://arduino-projekte.info/produkt/node-mcu-esp32-38pin/
THX

**Reply by henne49** [reply]
I have now installed the Binary, where do I set the MQTT Server?

**Reply by stromibaer** [reply]
In MqTT settings, the hostname:

![image](https://user-images.githubusercontent.com/47776842/202418512-7a6684ed-72bc-4558-8965-bcf5950161ec.png)

**Comment by henne49** [comment]
So I can only use Home Assistant or Victron? As I need both at the same time

**Reply by stromibaer** [reply]
So, you are running a Mosquitto server on the Home Assistant server?
I have similiar installation and add following lines in the Mosquitto default.conf file to forward the MqTT publishs from the Mosquitto server to the Victron Cerbo device:

/etc/mosquitto/conf.d/default.conf


# connection to Victron

connection victron
address 192.168.x.y:1883

topic # out 0
topic # in 0


So the Mosquitto server on the Home Assistant acts as bridge to the Victron Cerbo device.

**Reply by stromibaer** [reply]
https://github.com/freakent/dbus-mqtt-devices/ is already installed on the Victron Cerbo device?

With the dbus-mqtt-devices is a open task --> https://github.com/freakent/dbus-mqtt-devices/issues/23 
to adopt the services file to be able to configure the inverters side (AC-IN or AC-OUT).

**Reply by henne49** [reply]
I would really appreciate a documentation on how to use, then I can test. I'm still confused, how Home-assistant/Mosquitto knows what to forward to victron.

**Reply by henne49** [reply]
> https://github.com/freakent/dbus-mqtt-devices/ is already installed on the Victron Cerbo device?
> 
> With the dbus-mqtt-devices is a open task --> [freakent/dbus-mqtt-devices#23](https://github.com/freakent/dbus-mqtt-devices/issues/23) to adopt the services file to be able to configure the inverters side (AC-IN or AC-OUT).

ah no, my script does actually pull the API Data and inject in dbus. So my software is an alternative for dbus-mqtt-devices and can use dfault opendtu

**Reply by stromibaer** [reply]
> I would really appreciate a documentation on how to use, then I can test. I'm still confused, how Home-assistant/Mosquitto knows what to forward to victron.

With the additional lines in the config, the mosquitto on homeassistant will publish all topics out to the Victron Cerbo and all topics coming from Victron Cerbo to the mosquitto in, will be send to the subscribers.

**Reply by stromibaer** [reply]
I made a pull request: https://github.com/tbnobody/OpenDTU/pull/358 with some additional infos. Hoping the code comes in upstream and  i will work on documentation.

**Comment by magic21nrw** [comment]
@stromibaer Do you have a most recent version of this project actively running on Victron OS >3.00? I used the henne-script but with Version >3.00 it does not show the pvinverter anymore. 

I have seen you have a repository up and running but it seems that it never merged into openDTU because of some problems. You also said you deleted the repository because you messed something up. In your repositories is still a version visible - is this a new and running version? Maybe we could adapt it to the new Venus OS 3.XX and the most recent OpenDTU version.

Any thoughts?

PS: German is also fine with me...

**Comment by jetronic18s** [comment]
@stromibaer 
The idea of integrating Victron support directly into openDTU is great. I am excited to see if there is any news.

**Comment by Gelegenheitsbastler** [comment]
Eine direkte Einbindung in OpenDTU wäre eine tolle Sache. Ich habe zwar nicht viel Ahnung von Software, würde mich aber als Tester anbieten. 
Akutell: Venus OS auf Raspberry mit MPII 5k und diversen WR und MPPT. Leider sind meine beiden Hoymiles WR die "neuen" 800W-2T, so dass sie aktuell noch nicht von OpenDTU unterstützt werden. Bis das möglich ist, gehe ich den Umweg über Shelly EM, um sie im Victron Universum sehen zu können. Viel lieber würde ich das aber über die OpenDTU machen.

**Reply by henne49** [reply]
Unser script https://github.com/henne49/dbus-opendtu macht genau das möglich und es geht auch mit der aktuellen Victeon Version wieder.

**Reply by magic21nrw** [reply]
Meinst Du das Problem mit der Version >3.00 bei dem der PVInverter nicht mehr angezeigt wird und die Consumer nicht mehr richtig berechnet werden bzw. der PVinverter verschwunden ist?
Aktuell bin ich noch auf Version <3.0 und da läuft es ja ohne Probleme.

**Reply by henne49** [reply]
Ja genau, sollte in V3.13 von Venus OS laufen
<img width="517" alt="grafik" src="https://github.com/tbnobody/OpenDTU/assets/4662326/95ec9d5c-0aef-4cf5-9bcc-70fac1dc3eb5">

**Reply by Heinz7000** [reply]
Hallo,
bei mir wird der PVInverterleider in der Version 3.13 nicht angezeigt.
![image](https://github.com/tbnobody/OpenDTU/assets/129217773/296b2e1c-38fc-4dfd-98d4-339fcc33ab6f)
![image](https://github.com/tbnobody/OpenDTU/assets/129217773/3558c315-600f-4dfa-b00b-a5f910000bb7)

Muss ich da irgendwo noch etwas besonderes einstellen?

Lg

**Reply by henne49** [reply]
da ist aber auch keine Batterie? und kein Victron Inverter oder was sehe ich da falsch?

**Reply by Heinz7000** [reply]
Hallo,
nee, da hast du recht. Ist noch im Aufbau. Da kommt noch ein Multiplus 2 und eine Batterie. 

Lg

**Reply by henne49** [reply]
danach sollte das gehen

**Reply by Heinz7000** [reply]
Super, Danke!

Lg

**Reply by TxHxS** [reply]
Hi, ist schon bekannt ob das Script auf dem Cerbo mit v3.32 noch läuft ? Bei mir wird nichts mehr angezeigt. Nochmalige installation des scripts brachte keine Besserung :(

**Reply by henne49** [reply]
@TxHxS solche Fragen bitte in einem separatem Thread starten, zumindest bei mir läuft die latest version auf einem GX Device mit 3.32

**Reply by Gelegenheitsbastler** [reply]
Das kann ich bestätigen. Auch hier werden die in OpenDTU verbundenen Wechselrichter korrekt in der 3.32 angezeigt.

**Reply by TxHxS** [reply]
Danke, werde das Script nochmals neu installieren und testen. Sorry für den fehl-post hier
